python/ray/util/sgd/v2/backends/backend.py

```diff
@@ -55,20 +59,140 @@ def start(self, initialization_hook: Optional[Callable[[], None]] = None):
+        """Fetch next results produced by ``sgd.report()`` from each worker.
+
+        Assumes ``start_training`` has already been called.
+
+        Returns:
+            A list of dictionaries of values passed to ``sgd.report()`` from
+                each worker. Each item corresponds to an intermediate result
+                a single worker. If there are no more items to fetch,
+                returns None.
+        """
+
+        def get_next():
+            # Get the session for this worker.
+            try:
+                session = get_session()
+            except ValueError:
+                # Session is not initialized yet.
+                raise RuntimeError("`fetch_next_result` has been called "
+                                   "before `start_training`. Please call "
+                                   "`start_training` before "
+                                   "`fetch_next_result`.")
+
+            # Release the lock to trigger training to continue.
+            session.continue_lock.release()
+
+            result = None
+            # While training is still ongoing, attempt to get the result.
+            while result is None and session.training_thread.is_alive():
+                try:
+                    result = session.result_queue.get(
+                        block=True, timeout=RESULT_FETCH_TIMEOUT)
+                except queue.Empty:
+                    pass
+
+            # If no result was found, then the runner must no longer be alive.
+            if result is None:
+                # Try one last time to fetch results in case results were
+                # reported in between the time of the last check and the
+                # termination of the thread runner.
+                try:
+                    result = session.result_queue.get(
+                        block=True, timeout=RESULT_FETCH_TIMEOUT)
+                except queue.Empty:
+                    pass
+
+            # Return None if there are no more results to fetch.
+            return result
+
+        futures = self.worker_group.execute_async(get_next)
+        results = self.get_with_failure_handling(futures)
+
+        # Check if any worker returned None.
+        if any(r is None for r in results):
+            # Either all workers have results or none of them do.
+            if not all(r is None for r in results):
+                raise RuntimeError("Some workers returned results while "
+                                   "others didn't. Make sure that "
+                                   "`sgd.report()` is called the same number "
+                                   "of times on all workers.")
+            else:
+                results = None
+
+        return results
+
+    def finish_training(self) -> List[T]:
+        """Finish training and return final results. Propagate any exceptions.
+
+        Blocks until training is finished on all workers.
+
+        Assumes `start_training` has already been called.
```

Comment:
What happens if this is called before `fetch_next_result` returns `None`?

Reply:
Ah yeah this is a good point. I changed it so that `finish_training` can be called even if `fetch_next_result` has not been exhausted. This would useful for when we implement the generator API. A test is added for this as well.